design-proposals/monitoring-code-refactor/monitoring-code-refactor.md

```diff
@@ -0,0 +1,223 @@
+          help: Number of example operator pods in the cluster
+          type: gauge
+          expr: sum(up{namespace='$NAMESPACE', pod=~'example-operator-.*'}) or vector(0)
+      
+      alerts:
+        - name: OperatorDown
+          expr: example_operator_number_of_pods == 0
+          for: 5m
+          annotations:
+            summary: Example operator is down
+            description: Example operator is down for more than 5 minutes.
+            runbook_url: $RUNBOOK_BASE_URL/OperatorDown
+          labels:
+            severity: critical
+```
+
+But components can have a single file only for metrics or for a specific group
+of metrics, for example:
+
+```yaml
+observability:
+   groups:
+      - name: operator_resources
+        common_labels:
+           - name: grouped
+             value: this_is_grouped
+        metrics:
+           - name: example_operator_reconcile_count
+             help: Number of times the operator has executed the reconcile loop
+             type: counter
+             stability: stable
+             labels:
+                - name: another_label
+                  value: only_for_this_metric
+
+      - name: other_group
+        metrics:
+          ...
+```
+
+In the end, all the declarative configuration will be parsed and joined into a
+single `Observability` struct, which will be used to create the metrics, alerts,
+and recording rules. The same idea will be applied to other resources.
+
+It should be possible to:
+
+- create a single file with all groups of metrics, alerts, and recording rules
+
+- create multiple files, each with a specific set of groups. With each group
+containing its metrics, alerts, and recording rules
+
+- create multiple files, each with multiple groups with only one type of
+resource (metrics, alerts, or recording rules). For example, one file with all
+metrics, another with all alerts, and another with all recording rules.
+
+### Utilities
+
+The operator observability package will also have utilities to:
+- Generate the documentation for metrics, alerts, and recording rules
+- Lint metrics and recording rules names
+- Validate alerts configuration
+- Generate resource definitions PrometheusRule, ServiceMonitor, etc
+- Define allow-list, deny-list, and opt-in metrics
+
+## Action Plan
+
+### Phase 1 - Decouple monitoring logic from business logic
+
+1) Move the monitoring logic from the business logic code to a dedicated package
+in each component repository
+
+### Phase 2 - Use a generic operator-observability package
+
+1) Create the operator-observability package with the generic monitoring code
+   and utilities
+
+2) Document how to use the operator-observability package for adding new
+   metrics, alerts, and recording rules
+
+3) Update the monitoring code in each component to use the
+operator-observability package
+
+### Phase 3 - Have a declarative way to define metrics and alerts
+
+1) Update the operator-observability package to have a declarative way to
```

Comment:
I think we should keep the Prometheus way of creating the metrics and alert and not create another layer on top of it.

Reply:
The overhead of maintaining such a layer is not worth the effort IMHO